ccore/include/pyclustering/parallel/parallel.hpp

```diff
@@ -148,40 +212,36 @@ Advanced uses might use one of the define to use specific implementation of the
 @param[in] p_begin: initial iterator from that the loop starts.
 @param[in] p_end: end iterator that defines when the loop should stop `iter != p_end`.
 @param[in] p_task: body of the loop that defines actions that should be done for each element.
+@param[in] p_threads: amount of threads that are going to be used for processing (by default efficient amount of threads).
 
 */
 template <typename TypeIter, typename TypeAction>
-void parallel_for_each(const TypeIter p_begin, const TypeIter p_end, const TypeAction & p_task) {
+void parallel_for_each(const TypeIter p_begin, const TypeIter p_end, const TypeAction & p_task, const std::size_t p_threads = AMOUNT_THREADS) {
 #if defined(PARALLEL_IMPLEMENTATION_ASYNC_POOL)
-    const std::size_t step = std::distance(p_begin, p_end) / (AMOUNT_THREADS + 1);
+    const std::size_t interval_length = std::distance(p_begin, p_end);
+    const std::size_t step = std::max(interval_length / p_threads, std::size_t(1));
+
+    std::size_t amount_threads = static_cast<std::size_t>(interval_length / step);
+    if (amount_threads > p_threads) {
+        amount_threads = p_threads;
+    }
+    else if (amount_threads > 0) {
+        amount_threads--;   /* current thread is also considered. */
+    }
 
     auto current_start = p_begin;
     auto current_end = p_begin + step;
 
-    std::vector<std::size_t> captured_feature;
+    std::vector<std::future<void>> future_storage(amount_threads);
 

```

Comment:
- [ ] Possible bug. If  amount of threads is bigger than amount of data to process, then we could go out of range.

Reply:
Comment above.